### DISCUSSION TITLE
RealtimeChannel subscription status reports "SUBSCRIBED" then "CLOSED" immediately after

### DISCUSSION THREAD

**Discussion by scoover** [supabase/supabase#21410] [Questions]
I've created NextJS React custom hooks to interface with several tables with [channel subscriptions](https://supabase.com/docs/reference/javascript/subscribe?example=listen-to-a-specific-table). I'm confused because one of the hooks (2/3 work great) often reports a subscription `status` "SUBSCRIBED" immediately followed by "CLOSED".

```
.subscribe((status) => {
    if (["SUBSCRIBED", "CLOSED"].includes(status) && visible) {
        fetchPoints(`subscribe=${status}`);
    } else if (visible) {
        stopUsePoints(`subscribe=${status}`);
    }
```

In the code snippet, `fetchPoints()` fetches all relevant rows (but the subscription is broken once it reports "CLOSED") and `stopUsePoints()` calls `removeChannel()`.

I can't find documentation on what the "CLOSED" status means, how best to handle it, and why a subscription would close (apparently on its own).

My current hypothesis somehow the initial `fetchPoints()` operation (from the "SUBSCRIBED" event) is causing a fault or escalating an issue that is triggering the subsequent "CLOSED" subscription message? `fetchPoints()` looks like this and I can see that it is receiving valid values for `gameIds` ([228]) and `teamId` (2):

```
    const fetchPoints = async (status?: string) => {
      setIsLoading(true);
      const { data } = await supabase
        .from("points")
        .select()
        .in(
          gameIds.length ? "game_id" : "team_id",
          gameIds.length ? gameIds : [teamId]
        )
        .order("created_at", { ascending: true });
      setPoints(data);
      setIsLoading(false);
    };
```

Thanks for your help.

**Comment by GaryAustin1** [comment]
What do you do on CHANNEL_ERROR?
You should log out status.  
You should check realtime logs in the dashboard for errors.

Also is your goal to load initial data when you are connected to realtime?  If so then you really need to load your data something like this (the .on part)
```
 .subscribe((status) => {
                console.log('subscribe_status, ' + status);
                if (status !== 'SUBSCRIBED') {
                    connectionErrorHandler(status)
                }
            })
            .on('system', {}, payload => {                // need this to know when REALLY connected
                if (payload.extension === 'postgres_changes') {
                        console.log('postgres_changes received, load initial data')
                        handleTableInit(memoryTable,tableName,primaryCol)
                    }
            })
 ```
 https://github.com/supabase/realtime/issues/282#issuecomment-1287221414

**Comment by scoover** [comment]
Gary,

Thanks for your response on this. 

I tracked down the problem and it is sort of silly... I had duplicated code from a working custom hook to create a new one and had forgotten to change the channel name in my `supabase.channel("...")` call. When the two hooks were called in series, I think the "CLOSED" message was server closing the first caller with the same channel name so it could open up a new channel with the name. 🤦 

I had looked at a lot of your code and posts before I posted here— thank you for sharing.

One last quick question... I'm using the `.on('system', {}, payload =>` trigger that you recommend above, and, although it works great, my linter doesn't like it.

<img width="370" alt="image" src="https://github.com/supabase/supabase/assets/46952678/1037a4e4-987a-4f8d-bec1-8ae6b3a78ac0">

Do you have any suggestions on how to get the liner familiar with the "system" type? (It only recognizes "broadcast", "postgres_changes", and "presence".)

Thanks,
Randy

P.S. Actually... it's not just the linter. Vercel won't deploy as the issue ☝️ is triggering a build error.
<img width="1209" alt="image" src="https://github.com/supabase/supabase/assets/46952678/e7feac1a-f217-4117-a8c3-68f3b0eb3f78">

**Reply by GaryAustin1** [reply]
Sorry, no idea.
It is clearly "defined" in docs.  https://supabase.com/docs/guides/realtime/protocol#system-messages

**Reply by scoover** [reply]
Is there a chance I'm missing or using an outdated node dependency, @GaryAustin1?

```
    "@supabase/auth-helpers-nextjs": "^0.9.0",
    "@supabase/auth-ui-react": "^0.4.7",
    "@supabase/auth-ui-shared": "^0.1.8",
    "@supabase/realtime-js": "^2.9.3",
    "@supabase/supabase-js": "^2.39.7",
```

I'm particularly confused as it isn't just a linter/dev-environment issue but Vercel is also choking on building it. It feels like it is likely be a supabase-dependency version thing... but I just deleted `node_modules` and ran `npm install` from scratch and I'm seeing the same issue. 🤷

**Reply by scoover** [reply]
Here is my `package.json` for reference...

```
{
  "name": "nextjs",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@fortawesome/fontawesome-svg-core": "^6.5.1",
    "@fortawesome/free-brands-svg-icons": "^6.5.1",
    "@fortawesome/free-regular-svg-icons": "^6.5.1",
    "@fortawesome/free-solid-svg-icons": "^6.5.1",
    "@fortawesome/react-fontawesome": "^0.2.0",
    "@nivo/core": "^0.84.0",
    "@nivo/heatmap": "^0.84.0",
    "@nivo/pie": "^0.84.0",
    "@supabase/auth-helpers-nextjs": "^0.9.0",
    "@supabase/auth-ui-react": "^0.4.7",
    "@supabase/auth-ui-shared": "^0.1.8",
    "@supabase/realtime-js": "^2.9.3",
    "@supabase/supabase-js": "^2.39.7",
    "bootstrap": "^5.3.3",
    "date-fns": "^3.3.1",
    "eslint": "8.57.0",
    "eslint-config-next": "^14.1.0",
    "next": "^14.1.0",
    "react": "^18.2.0",
    "react-bootstrap": "^2.10.1",
    "react-dom": "^18.2.0",
    "react-geolocated": "^4.1.2"
  },
  "devDependencies": {
    "encoding": "^0.1.13"
  }
}
```

**Reply by scoover** [reply]
And here is the full channel configuration chain for reference...
```
        gamesListener.current = supabase
          .channel("public:games")
          .on(
            "postgres_changes",
            {
              event: "INSERT",
              schema: "public",
              table: "games",
              filter: `team_id=eq.${teamId}`,
            },
            (payload) => {
              insertGameInGames(payload.new as game);
            }
          )
          .on(
            "postgres_changes",
            {
              event: "DELETE",
              schema: "public",
              table: "games",
              filter: `team_id=eq.${teamId}`,
            },
            (payload) => {
              deleteGameInGames(payload.old as game);
            }
          )
          .on(
            "postgres_changes",
            {
              event: "UPDATE",
              schema: "public",
              table: "games",
              filter: `team_id=eq.${teamId}`,
            },
            (payload) => {
              updateGameInGames(payload.new as game);
            }
          )
          .on("system", {}, (payload) => {
            if (payload.extension === "postgres_changes") {
              fetchGames(`system=${payload.extension}`);
            }
          })          
          .subscribe((status) => {
            if (["SUBSCRIBED", "CLOSED"].includes(status) && visible) {
              logHook(`useGames status: ${status}`);
            } else if (visible) {
              stopUseGames(`subscribe=${status}`);
            }
          });            
```

You can see that "system" is just not an expected value.
<img width="228" alt="image" src="https://github.com/supabase/supabase/assets/46952678/66191ee4-ad6a-4228-a381-bc8e1048b773">

**Reply by GaryAustin1** [reply]
If you don't get someone else jumping in, you may want refocus with a new question.  I just use supabase-js and I linked you both the employees comment showing the test and the reference to that actually being a valid message.  I use the code fine in browser javascript.

I don't know anything about linting it or why Vercel would barf on it.  Sorry.

**Reply by GaryAustin1** [reply]
Or add a realtime-js issue.

**Reply by scoover** [reply]
Thanks, @GaryAustin1. I'm grateful for your guidance and responsiveness.

**Reply by scoover** [reply]
FYI @GaryAustin1... I revisited the link you provided ☝️ and found [this comment](https://github.com/supabase/realtime/issues/282#issuecomment-1617719644) from [NiklasPor](https://github.com/NiklasPor):
> It would be very nice if we could add the system type to the TypeScript definitions of realtime, so that we don't have to cast everything as any.
>...
>`.on('system' as any, {} as any, (payload: any) => {`

I added the typecasts and the linter warning resolved and the project compiled on Vercel... so maybe this linter/Typescript issue has been around for a while... 🎉 

(No need to respond to this, @GaryAustin1.)
